connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java

```diff
@@ -124,7 +126,19 @@ public void poll(long timeout) {
 
         do {
             if (coordinatorUnknown()) {
-                ensureCoordinatorReady(time.timer(Long.MAX_VALUE));
+                log.debug("Broker coordinator is marked unknown. Attempting discovery with a timeout of {}ms",
+                        coordinatorDiscoveryTimeoutMs);
+                if (ensureCoordinatorReady(time.timer(coordinatorDiscoveryTimeoutMs))) {
+                    log.debug("Broker coordinator is ready");
+                } else {
+                    log.debug("Can not connect to broker coordinator");
+                    if (assignmentSnapshot != null && !assignmentSnapshot.failed()) {
+                        log.info("Broker coordinator was unreachable for {}ms. Revoking previous assignment {} to " +
+                                "avoid running tasks while not being a member the group", coordinatorDiscoveryTimeoutMs, assignmentSnapshot);
+                        listener.onRevoked(assignmentSnapshot.leader(), assignmentSnapshot.connectors(), assignmentSnapshot.tasks());
+                        assignmentSnapshot = null;
```

Comment:
Thanks for improving the atomicity of clearing out the snapshot. However, IIUC the way this class is access by multiple threads means that the various places where `assignmentSnapshot` is read, the read may not be consistent and might result in a NPE. This line is the only place where `assignmentSnapshot` can be assigned to null after being not null, but that means other places where `assignmentSnapshot` is read could potentially produce a NPE. For example, `onJoinComplete(...)` (lines 197-199) does the following:
```
            if (assignmentSnapshot != null) {
                assignmentSnapshot.connectors().removeAll(newAssignment.revokedConnectors());
                assignmentSnapshot.tasks().removeAll(newAssignment.revokedTasks());
```
but if the broker coordinator is lost after the first line but before the second or third lines, the latter would produce a NPE. I think there are a few such places, and it'd be good to make the `assignmentSnapshot` member volatile (to ensure proper read semantics for multiple threads) and to use local variables to ensure consistent reads of the snapshot.

Reply:
Valid points. To be on the safe side I've made `assignmentSnapshot` volatile, and wherever it's used I'm using a local copy of the reference instead (`final ExtendedAssignment localAssignmentSnapshot = assignmentSnapshot`)

Testing with faults while this PR builds.